lib/ansible/modules/identity/cyberark/cyberark_authentication.py

```diff
@@ -0,0 +1,301 @@
+    headers = {'Content-Type': 'application/json'}
+    payload = ""
+
+    if state == "present":  # Logon Action
+
+        # Different end_points based on the use of shared logon authentication
+        if use_shared_logon_authentication:
+
+            end_point = "/PasswordVault/WebServices/auth/Shared/RestfulAuthenticationService.svc/Logon"
+
+        else:
+
+            end_point = "/PasswordVault/WebServices/auth/Cyberark/CyberArkAuthenticationService.svc/Logon"
+
+            # The payload will contain username, password
+            # and optionally use_radius_authentication and new_password
+            payload_dict = {"username": username, "password": password}
+
+            if use_radius_authentication:
+                payload_dict["useRadiusAuthentication"] = use_radius_authentication
+
+            if new_password is not None:
+                payload_dict["newPassword"] = new_password
+
+            payload = json.dumps(payload_dict)
+
+    else:  # Logoff Action
+
+        # Get values from cyberark_session already established
+        api_base_url = cyberark_session["api_base_url"]
+        validate_certs = cyberark_session["validate_certs"]
+        use_shared_logon_authentication = cyberark_session[
+            "use_shared_logon_authentication"]
+        headers["Authorization"] = cyberark_session["token"]
+
+        # Different end_points based on the use of shared logon authentication
+        if use_shared_logon_authentication:
+            end_point = "/PasswordVault/WebServices/auth/Shared/RestfulAuthenticationService.svc/Logoff"
+        else:
+            end_point = "/PasswordVault/WebServices/auth/Cyberark/CyberArkAuthenticationService.svc/Logoff"
+
+    result = None
+    changed = False
+    response = None
+
+    try:
+
+        response = open_url(
+            api_base_url + end_point,
+            method="POST",
+            headers=headers,
+            data=payload,
+            validate_certs=validate_certs)
+
+    except httplib.HTTPException as e:
+
+        module.fail_json(
+            msg="end_point=" +
+            api_base_url +
+            end_point,
+            payload=payload,
+            headers=headers,
+            exception=to_text(e),
+            status_code=e.code)
+
+    except Exception as e:
+
+        module.fail_json(
+            msg="end_point=" +
+            api_base_url +
+            end_point,
+            payload=payload,
+            headers=headers,
+            exception=to_text(e),
+            status_code=-1)
+
+    if response.getcode() == 200:  # Success
+
+        if state == "present":  # Logon Action
+
+            # Result token from REST Api uses a different key based
+            # the use of shared logon authentication
+            if use_shared_logon_authentication:
+                token = json.loads(response.read())["LogonResult"]
+            else:
+                token = json.loads(response.read())["CyberArkLogonResult"]
```

Comment:
Is there the potential that the response could be 200, but the JSON not include these keys?  Maybe they should be guarded with try/except to avoid exceptions.